From 335f9af5c4b3750a0994eac69e56a296903a6525 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 4 May 2026 22:47:16 -0700
Subject: [PATCH 7/8] Fix confused deputy in Bluetooth settings dashboard

The Bluetooth dashboard incorrectly trusted any caller from the Settings
package itself. On multi-pane devices, a deep link trampoline causes
external requests to appear as if they came from Settings.

This change uses getInitialCallingPackage() to correctly identify the
original caller and only enable AlwaysDiscoverable mode for trusted
system callers.

Bug: 471797085
Test: atest BluetoothDashboardFragmentTest BluetoothDashboardFragmentVulnerabilityTest
Flag: EXEMPT BUGFIX

Note: Autogenerated by MendIt (go/androidmendit)
NO_IFTTT=The changes are internal security checks in the fragment and do not affect the Catalyst screen structure or API.
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:0ca0778b89215e64f8573cb0487efd0b702721cf
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:47dca2d978c3eefb8725976196a907527f5f4928
Merged-In: I58b0c0898408562c1c088f14e0f3fff6ceaff1eb
Change-Id: I58b0c0898408562c1c088f14e0f3fff6ceaff1eb
---
 .../bluetooth/DevicePickerFragment.java       | 19 ++++++-
 .../BluetoothDashboardFragment.java           | 19 +++++--
 ...othDashboardFragmentVulnerabilityTest.java | 52 +++++++++++++++++++
 3 files changed, 84 insertions(+), 6 deletions(-)
 create mode 100644 tests/robotests/src/com/android/settings/connecteddevice/BluetoothDashboardFragmentVulnerabilityTest.java

diff --git a/src/com/android/settings/bluetooth/DevicePickerFragment.java b/src/com/android/settings/bluetooth/DevicePickerFragment.java
index 56bc5215b19..4b11b72beda 100644
--- a/src/com/android/settings/bluetooth/DevicePickerFragment.java
+++ b/src/com/android/settings/bluetooth/DevicePickerFragment.java
@@ -19,6 +19,7 @@ package com.android.settings.bluetooth;
 import static android.os.UserManager.DISALLOW_CONFIG_BLUETOOTH;
 
 import android.Manifest;
+import android.app.Activity;
 import android.app.settings.SettingsEnums;
 import android.bluetooth.BluetoothAdapter;
 import android.bluetooth.BluetoothDevice;
@@ -33,9 +34,11 @@ import android.view.Menu;
 import android.view.MenuInflater;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import androidx.annotation.VisibleForTesting;
 
 import com.android.settings.R;
+import com.android.settings.SettingsActivity;
 import com.android.settings.password.PasswordUtils;
 import com.android.settingslib.bluetooth.CachedBluetoothDevice;
 import com.android.settingslib.core.AbstractPreferenceController;
@@ -97,8 +100,7 @@ public final class DevicePickerFragment extends DeviceListPreferenceFragment {
         getActivity().setTitle(getString(R.string.device_picker));
         UserManager um = (UserManager) getSystemService(Context.USER_SERVICE);
         mScanAllowed = !um.hasUserRestriction(DISALLOW_CONFIG_BLUETOOTH);
-        mCallingAppPackageName = PasswordUtils.getCallingAppPackageName(
-                getActivity().getActivityToken());
+        mCallingAppPackageName = resolveInitialCallingPackage();
         if (!TextUtils.equals(mCallingAppPackageName, mLaunchPackage)) {
             Log.w(TAG, "sendDevicePickedIntent() launch package name is not equivalent to"
                     + " calling package name!");
@@ -107,6 +109,19 @@ public final class DevicePickerFragment extends DeviceListPreferenceFragment {
         setHasOptionsMenu(true);
     }
 
+    @Nullable
+    private String resolveInitialCallingPackage() {
+        Activity activity = getActivity();
+        if (activity instanceof SettingsActivity) {
+            return ((SettingsActivity) activity).getInitialCallingPackage();
+        }
+
+        if (activity != null) {
+            return PasswordUtils.getCallingAppPackageName(activity.getActivityToken());
+        }
+        return null;
+    }
+
     @Override
     public void onAttach(@NonNull Context context) {
         super.onAttach(context);
diff --git a/src/com/android/settings/connecteddevice/BluetoothDashboardFragment.java b/src/com/android/settings/connecteddevice/BluetoothDashboardFragment.java
index f66fb375507..839c1c36f5a 100755
--- a/src/com/android/settings/connecteddevice/BluetoothDashboardFragment.java
+++ b/src/com/android/settings/connecteddevice/BluetoothDashboardFragment.java
@@ -225,8 +225,7 @@ public class BluetoothDashboardFragment extends DashboardFragment {
         if (isCatalystEnabled()) {
             return;
         }
-        String callingAppPackageName = PasswordUtils.getCallingAppPackageName(
-                getActivity().getActivityToken());
+        String callingAppPackageName = resolveInitialCallingPackage();
         String action = getIntent() != null ? getIntent().getAction() : "";
         if (DEBUG) {
             Log.d(TAG, "onActivityCreated() calling package name is : " + callingAppPackageName
@@ -251,8 +250,7 @@ public class BluetoothDashboardFragment extends DashboardFragment {
         super.onStart();
         if (isCatalystEnabled()) {
             Activity activity = requireActivity();
-            String callingAppPackageName = PasswordUtils.getCallingAppPackageName(
-                    activity.getActivityToken());
+            String callingAppPackageName = resolveInitialCallingPackage();
             Intent intent = activity.getIntent();
             String action = intent != null ? intent.getAction() : "";
             if (DEBUG) {
@@ -275,6 +273,19 @@ public class BluetoothDashboardFragment extends DashboardFragment {
         }
     }
 
+    @Nullable
+    private String resolveInitialCallingPackage() {
+        Activity activity = getActivity();
+        if (activity instanceof SettingsActivity) {
+            return ((SettingsActivity) activity).getInitialCallingPackage();
+        }
+
+        if (activity != null) {
+            return PasswordUtils.getCallingAppPackageName(activity.getActivityToken());
+        }
+        return null;
+    }
+
     @VisibleForTesting
     boolean isAlwaysDiscoverable(@Nullable String callingAppPackageName, @Nullable String action) {
         return TextUtils.equals(SLICE_ACTION, action) ? false
diff --git a/tests/robotests/src/com/android/settings/connecteddevice/BluetoothDashboardFragmentVulnerabilityTest.java b/tests/robotests/src/com/android/settings/connecteddevice/BluetoothDashboardFragmentVulnerabilityTest.java
new file mode 100644
index 00000000000..1f42fc4a949
--- /dev/null
+++ b/tests/robotests/src/com/android/settings/connecteddevice/BluetoothDashboardFragmentVulnerabilityTest.java
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.settings.connecteddevice;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class BluetoothDashboardFragmentVulnerabilityTest {
+    private static final String SETTINGS_PACKAGE_NAME = "com.android.settings";
+    private static final String EXTERNAL_APP_PACKAGE_NAME = "com.external.app";
+    private static final String TEST_ACTION = "android.settings.BLUETOOTH_DASHBOARD_SETTINGS";
+
+    private BluetoothDashboardFragment mFragment;
+
+    @Before
+    public void setUp() {
+        mFragment = new BluetoothDashboardFragment();
+    }
+
+    @Test
+    public void isAlwaysDiscoverable_initialPackageIsSettings_returnsTrue() {
+        // This simulates the case where resolveInitialCallingPackage() returns Settings.
+        // It should still return true because Settings is trusted.
+        assertThat(mFragment.isAlwaysDiscoverable(SETTINGS_PACKAGE_NAME, TEST_ACTION)).isTrue();
+    }
+
+    @Test
+    public void isAlwaysDiscoverable_initialPackageIsExternal_returnsFalse() {
+        // This simulates the case where resolveInitialCallingPackage() returns the external app.
+        // It should return false because the external app is not trusted.
+        assertThat(mFragment.isAlwaysDiscoverable(EXTERNAL_APP_PACKAGE_NAME, TEST_ACTION))
+                .isFalse();
+    }
+}
-- 
2.55.0

